README.md
[![GitHub package.json version](https://img.shields.io/github/package-json/v/xpack/update-checker-ts)](https://github.com/xpack/update-checker-ts/blob/mater/package.json)
[![npm (scoped)](https://img.shields.io/npm/v/@xpack/update-checker.svg)](https://www.npmjs.com/package/@xpack/update-checker)
[![license](https://img.shields.io/github/license/xpack/update-checker-js.svg)](https://github.com/xpack/update-checker-js/blob/xpack/LICENSE)

# A Node.js ES6 module to notify when a newer version of a npm package is available

This project provides a **TypeScript** Node.js **ES6** module with
a class to check if a more recent version of a npm package is available.

The open source project is hosted on GitHub as
[xpack/update-checker-ts](https://github.com/xpack/update-checker-ts).

## Maintainer & developer info

This page documents how to use this module in an user application.
For maintainer information, see the separate
[README-MAINTAINER](https://github.com/xpack/update-checker-ts/blob/master/README-MAINTAINER.md)
page.

## Prerequisites

A recent [Node.js](https://nodejs.org) (>=16.0.0), since the TypeScript code
is compiled into ECMAScript 2020 code with ES6 modules.

## Install

The module is available as
[`@xpack/update-checker`](https://www.npmjs.com/package/@xpack/update-checker/)
from the public [`npmjs`](https://www.npmjs.com) repository;
it can be added as a dependency to any TypeScript or JavaScript
project with `npm install`:

```console
npm install --save @xpack/update-checker@latest
```

The module does not provide any executables, and generally there are no
reasons to install it globally.

## User info

This section is intended for those who plan to use this module in their
own projects.

The `@xpack/update-checker` module can be imported into both TypeScript
and JavaScript ES6 Node.js code with:

```typescript
import { UpdateChecker } from '@xpack/update-checker'
```

Note: as per the ES6 specs, importing ES6 modules in legacy
CommonJS modules is no longer possible.

The module can be included in any application and the class can be used
directly or as a base class for a derived class to implement a custom
behaviour (like a custom notifier).

A typical use case is as part of a CLI application.

Example:

```javascript
import { UpdateChecker } from '@xpack/update-checker'
import { Logger } from '@xpack/logger'

async main(argv) {
  const log = new Logger({ level: 'info' })

  // ... Other initializations.

  // Read in the current package JSON (bring your own function here).
  const packageJson = await readPackageJson()

  // Create on instance of notifier class, configured for the
  // current package.
  const updateChecker = new UpdateChecker({
    log: log,
    packageName: packageJson.name,
    packageVersion: packageJson.version
  })

  // Start the update checker, as an asynchronous function
  // running in parallel.
  await updateChecker.initiateVersionRetrieval()

  // ... Perform the application actual work.

  // Before returning, possibly send a notification to the console.
  await updateChecker.notifyIfUpdateIsAvailable()

  return exitCode
}
```

The check is no longer performed during the next 24 hours; this duration
can be configured via the `checkUpdatesIntervalSeconds` constructor property.

The check is skipped if one of the following conditions are met:

- running in a CI environment
- stdout is not a TTY
- the process is running as root
- the program is installed in system locations
- the `NO_<packageName>_UPDATE_NOTIFIER` variable is present in the
  environment map (configurable)

### Reference

For more details on the available class definitions, including all methods,
accessors, properties, etc,
please see the TypeDoc
[reference pages](https://xpack.github.io/update-checker-ts).

## Known problems

- none

## Status

The `@xpack/update-checker` module is fully functional and stable.

The main client for this module is the `xpm` CLI application.

## Tests

The module is tested
with 100% coverage and CI tested on every push via GitHub
[Actions](https://github.com/xpack/update-checker-ts/actions).

## Change log - incompatible changes

According to [semver](https://semver.org) rules:

> Major version X (X.y.z | X > 0) MUST be incremented if any
backwards incompatible changes are introduced to the public API.

### v2.0.0

The project was migrated to TypeScript and the code is now compiled into
**ES6** modules, that can be consumed by both TypeScript and JavaScript
packages. The minimum node engine was updated to >16.0.0.

Other changes:

- the test environment was moved outside the constructor
- the parameter of `sendNotification()` was removed in favour of the
  instance property
- the name of the environment variable was added to the constructor

## License

The original content is released under the
[MIT License](https://opensource.org/license/mit/),
with all rights reserved to
[Liviu Ionescu](https://github.com/ilg-ul).


CHANGELOG.md
# Change log

Changes in reverse chronological order.

Refer to GitHub [issues](https://github.com/xpack/update-checker-js/issues).

## 2023-03-18

* 08ba4d3 READMEs updates
* 02010ff workflows renames
* 7e57d17 package-lock.json update
* 2d46f88 package.json cosmetics
* 8ecad1a import cosmetics
* f353b6f update licenses
* b5ed29d eslint max-len

## 2023-03-09

* 97502f8 010-constructor.ts: remove *ConstructorParameters
* 4796da9 update-checker.ts: move constructor type in place
* 42d8f2c update-checker.ts: update typedoc
* c367650 README update
* 4dceb9c typedoc.json: cosmetics
* a6171ae README update
* 84a6855 020-checker.ts: update for new variable name
* 89aa78e #1: make the environment variable configurable
* 61f45bd #22: default variable includes the package name

## 2023-03-07

* ac004c3 .github/workflows update actions
* b6ca962 bump node 16
* d3a0c92 update-checker.ts: re-place c8 ignore start

## 2023-03-05

* 90dbd85 update-checker.ts: rework istanbul ignore
* 3773b33 update-checker.ts: rework c8 ignore start
* 8848fd9 typedoc.yml: use node 14
* 37e3ce4 update-checker.ts: rework c8 ignore start
* 785a73d READMEs updates
* 7590f2b CHANGELOG update
* 8b235d9 .github/workflows update
* a59d6a5 .vscode/settings.json
* 8f27aff README update
* da27147 READMEs updates
* d12f2d0 .vscode/settings.json
* a45bd2a update-checker.ts: cosmetics
* ff9412d typedoc.json: add sort
* 8296491 add more typedoc metadata
* b4e0248 tests: use PublicUpdateChecker
* 1f01dcc update-checker.ts: add /* c8 ignore * */
* 0221e9d update-checker.ts: naming cosmetics
* 3666c9b update-checker.ts: more protected
* 1978e65 update-checker.ts: this.* in sendNotifications

## 2023-03-04

* 2c479d9 update-checker.ts: typedoc
* 09c0cd5 typedoc.json: url comment
* b6b174d typedoc.json: intentionallyNotExported
* f817a04 package.json: typedoc Verbose
* 1453d6d update-checker.ts: fix NO_NPM_UPDATE_NOTIFIER test
* bc312ae update-checker.ts: fix dirname
* b524a21 mock-console.ts: remove stderrStream
* f85c92d READMEs updates
* 37fc669 add & use separate UpdateCheckerTestEnvironment
* 1c20bf0 package-lock.json: update
* 59de4bf package.json: add eslintConfig
* 668b8fe remove eslint comments from sources
* 267df9d configure esm only builds
* 1be8aab typedoc: add custom css

## 2023-03-03

* f28d899 package-lock.json update
* 146f9e0 mock-logger.ts: remove constructor
* f6c05f5 READMEs updates
* 6c4ef6f add typedoc.json
* b6add7f .vscode/settings.json: ignoreWords
* c09b383 update for .ts

## 2023-03-02

* 2382daa rename .ts, move to src, tests, shorten copyright
* c65bd04 update-checker.js: reorder require()
* 42f3968 tests/020-checker.js: fix standard issues
* ddc65a3 package.json: bump deps

## 2021-07-20

* publish v1.2.0
* [#16] - process errors while retrieving latest version
* [#15] - deprecate reference to package user-home
* [#14] - update to package make-dir
* [#13] - use fs promises

## 2020-12-11

* publish v1.1.2
* bump dependencies

## 2019-11-25

* publish v1.1.1
* tweak message, update silently ignored
* switch to GitHub Actions
* enforce 100% coverage

## 2019-11-11

* publish v1.1.0
* [#2] Test environment as a map not an array
* fix standard issues (hasOwnProperty)
* bump deps

## 2019-01-25

* publish v1.0.0
* initial version, extracted from @ilg/cli-start-options-js.
* reworked for testability
* tests added, 100% coverage


test/common.js
/*
 * This file is part of the xPack distribution
 *   (http://xpack.github.io).
 * Copyright (c) 2019 Liviu Ionescu.
 *
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without
 * restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom
 * the Software is furnished to do so, subject to the following
 * conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */

'use strict'
/* eslint valid-jsdoc: "error" */
/* eslint max-len: [ "error", 80, { "ignoreUrls": true } ] */

// ----------------------------------------------------------------------------

const util = require('util')

// ----------------------------------------------------------------------------

class MockLog {
  constructor () {
    this.lines = []
    this.isDebug = false
  }

  warning (msg = '', ...args) {
    const str = util.format(msg, ...args)
    this.lines.push('warning: ' + str)
  }

  info (msg = '', ...args) {
    const str = util.format(msg, ...args)
    this.lines.push(str)
  }

  debug (msg = '', ...args) {
    const str = util.format(msg, ...args)
    this.lines.push(str)
  }

  trace (msg = '', ...args) {
    const str = util.format(msg, ...args)
    this.lines.push(str)
  }

  clear () {
    this.lines = []
  }
}

// ----------------------------------------------------------------------------
// Node.js specific export definitions.

// By default, `module.exports = {}`.
// The Main class is added as a property to this object.

module.exports.MockLog = MockLog

// In ES6, it would be:
// export class MockLog { ... }
// ...
// import { MockLog } from 'common.js'

// ----------------------------------------------------------------------------


tests/common.ts
/*
 * This file is part of the xPack project (http://xpack.github.io).
 * Copyright (c) 2019 Liviu Ionescu. All rights reserved.
 *
 * Permission to use, copy, modify, and/or distribute this software
 * for any purpose is hereby granted, under the terms of the MIT license.
 *
 * If a copy of the license was not distributed with this file, it can
 * be obtained from https://opensource.org/licenses/MIT/.
 */

'use strict'
/* eslint valid-jsdoc: "error" */
/* eslint max-len: [ "error", 80, { "ignoreUrls": true } ] */

// ----------------------------------------------------------------------------

const util = require('util')

// ----------------------------------------------------------------------------

class MockLog {
  constructor () {
    this.lines = []
    this.isDebug = false
  }

  warning (msg = '', ...args) {
    const str = util.format(msg, ...args)
    this.lines.push('warning: ' + str)
  }

  info (msg = '', ...args) {
    const str = util.format(msg, ...args)
    this.lines.push(str)
  }

  debug (msg = '', ...args) {
    const str = util.format(msg, ...args)
    this.lines.push(str)
  }

  trace (msg = '', ...args) {
    const str = util.format(msg, ...args)
    this.lines.push(str)
  }

  clear () {
    this.lines = []
  }
}

// ----------------------------------------------------------------------------
// Node.js specific export definitions.

// By default, `module.exports = {}`.
// The Main class is added as a property to this object.

module.exports.MockLog = MockLog

// In ES6, it would be:
// export class MockLog { ... }
// ...
// import { MockLog } from 'common.js'

// ----------------------------------------------------------------------------


index.js
/*
 * This file is part of the xPack distribution
 *   (http://xpack.github.io).
 * Copyright (c) 2019 Liviu Ionescu.
 *
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without
 * restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom
 * the Software is furnished to do so, subject to the following
 * conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */

'use strict'
/* eslint valid-jsdoc: "error" */
/* eslint max-len: [ "error", 80, { "ignoreUrls": true } ] */

// ----------------------------------------------------------------------------

/**
 * This is the module entry point, the file that is processed when
 * `require('@xpack/update-checker')` is called.
 *
 * For this to work, it must be linked from `package.json` as
 * `"main": "./index.js",`, which is, BTW, the default behaviour.
 *
 * This file does not define the classes itself, but imports them
 * from various implementation files, and re-exports them.
 *
 * To import classes from this module into Node.js applications, use:
 *
 * ```javascript
 * const UpdateChecker = require('@xpack/update-checker').UpdateChecker
 * ```
 */

const UpdateChecker = require('./lib/update-checker.js').UpdateChecker

// ----------------------------------------------------------------------------
// Node.js specific export definitions.

// By default, `module.exports = {}`.
// The Main class is added as a property with the same name to this object.

module.exports.UpdateChecker = UpdateChecker

// In ES6, it would be:
// export class UpdateChecker { ... }
// ...
// import { UpdateChecker } from 'update-checker.js'

// ----------------------------------------------------------------------------


lib/update-checker.js
/*
 * This file is part of the xPack distribution
 *   (http://xpack.github.io).
 * Copyright (c) 2019 Liviu Ionescu.
 *
 * Permission is hereby granted, free of charge, to any person
 * obtaining a copy of this software and associated documentation
 * files (the "Software"), to deal in the Software without
 * restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom
 * the Software is furnished to do so, subject to the following
 * conditions:
 *
 * The above copyright notice and this permission notice shall be
 * included in all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES
 * OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
 * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
 * WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 */

'use strict'
/* eslint valid-jsdoc: "error" */
/* eslint max-len: [ "error", 80, { "ignoreUrls": true } ] */

// ----------------------------------------------------------------------------

// https://nodejs.org/docs/latest-v10.x/api/
const assert = require('assert')
const os = require('os')
const path = require('path')
const fs = require('fs')

// ----------------------------------------------------------------------------

const fsPromises = fs.promises

// ----------------------------------------------------------------------------

// https://www.npmjs.com/package/del
const del = require('del')

// https://www.npmjs.com/package/is-ci
// Used in place.
// const isCI = require('is-ci')

// https://www.npmjs.com/package/is-installed-globally
// Used in place.
// const isInstalledGlobally = require('is-installed-globally')

// https://www.npmjs.com/package/is-path-inside
// Note v4 no longer supports `require()`, but `import`.
const isPathInside = require('is-path-inside')

// https://www.npmjs.com/package/make-dir
const makeDir = require('make-dir')

// https://www.npmjs.com/package/latest-version
const latestVersionPromise = require('latest-version')

// https://www.npmjs.com/package/semver
const semver = require('semver')

// ----------------------------------------------------------------------------

// Modules that return a Boolean. Used in place.

// Default variables.
const timestampsFolderAbsolutePath = path.join(os.homedir(), '.config',
  'npm-timestamps')
const timestampSuffix = '-update-check'
const defaultCheckUpdatesIntervalSeconds = 24 * 60 * 60

// ============================================================================

// export
class UpdateChecker {
  // --------------------------------------------------------------------------

  // Constructor: use parent definition.
  constructor (params) {
    assert(params)

    assert(params.log)
    this.log = params.log

    const log = this.log
    log.trace(`${this.constructor.name}.constructor()`)

    assert(params.packageName)
    this.packageName = params.packageName

    assert(params.packageVersion)
    this.packageVersion = params.packageVersion

    // Optional parameters.
    this.timestampsFolderAbsolutePath = params.timestampsFolderAbsolutePath ||
      timestampsFolderAbsolutePath

    // Uses the folder path.
    this.timestampFileAbsolutePath = path.join(
      this.timestampsFolderAbsolutePath,
      this.packageName + timestampSuffix
    )

    this.checkUpdatesIntervalMilliseconds =
      (params.checkUpdatesIntervalSeconds ||
        defaultCheckUpdatesIntervalSeconds) * 1000

    this.private_ = {}

    // For testability reasons, allow to override some conditions.
    // This may be also used for special environments and use cases.
    this.private_.processEnv = params.env || process.env || {}

    if (Object.prototype.hasOwnProperty.call(params, 'isCI')) {
      this.private_.isCI = params.isCI
    } else {
      this.private_.isCI = require('is-ci')
    }

    if (Object.prototype.hasOwnProperty.call(params, 'isTTY')) {
      this.private_.isTTY = params.isTTY
    } else {
      this.private_.isTTY = process.stdout.isTTY
    }

    if (Object.prototype.hasOwnProperty.call(params, 'isRunningAsRoot')) {
      this.private_.isRunningAsRoot = params.isRunningAsRoot
    } else {
      /* istanbul ignore next */
      if (os.platform() !== 'win32') {
        this.private_.isRunningAsRoot =
          process.geteuid && process.geteuid() !== process.getuid()
      } else {
        /* istanbul ignore next */
        this.private_.isRunningAsRoot = false
      }
    }

    if (Object.prototype.hasOwnProperty.call(params, 'isInstalledGlobally')) {
      this.private_.isInstalledGlobally = params.isInstalledGlobally
    } else {
      this.private_.isInstalledGlobally = require('is-installed-globally')
    }

    if (Object.prototype.hasOwnProperty.call(params, 'isInstalledAsRoot')) {
      this.private_.isInstalledAsRoot = params.isInstalledAsRoot
    } else {
      this.private_.isInstalledAsRoot = false
      /* istanbul ignore next */
      if (os.platform() !== 'win32') {
        if (this.private_.isInstalledGlobally &&
          isPathInside(__dirname, '/usr/local')) {
          // May not be very reliable if installed in another system location.
          this.private_.isInstalledAsRoot = false
        }
      }
    }
  }

  /**
   * @summary Initiate a procedure to retrieve the version.
   *
   * @returns {undefined} Nothing.
   *
   * @description
   * Since this is a potentially long operation, it is initiated at the
   * beginning and executed at the end, giving it a chance to run in
   * parallel.
   *
   * The operation is not started when running in a CI environment,
   * when not running from a console, or the environment variable
   * `NO_NPM_UPDATE_NOTIFIER` is defined.
   */
  async initiateVersionRetrieval () {
    const log = this.log
    log.trace(`${this.constructor.name}.initiateVersionRetrieval()`)

    this.latestVersionPromise = undefined

    const envName = 'NO_NPM_UPDATE_NOTIFIER'
    if (this.checkUpdatesIntervalMilliseconds === 0 ||
      this.private_.isCI ||
      !this.private_.isTTY ||
      Object.prototype.hasOwnProperty.call(this.private_.processEnv, envName)) {
      log.trace(`${this.constructor.name}:` +
        ' do not fetch latest version number.')
      return
    }

    if (await this.didIntervalExpire_(this.checkUpdatesIntervalMilliseconds)) {
      log.trace(`${this.constructor.name}:` +
        ' fetching latest version number...')

      // Create the promise, do not wait for it yet.
      this.latestVersionPromise = this.prepareLatestVersionPromise()
    }
  }

  /**
   * @summary Create the promise to get the latest version.
   *
   * @returns {Promise} A promise to retrieve the version.
   */
  prepareLatestVersionPromise () {
    return latestVersionPromise(this.packageName)
  }

  /**
   * @summary Compare versions and notify if an update is available.
   *
   * @returns {undefined} Nothing.
   *
   * @description
   * Await for the actual version to be retrieved, compare with the
   * current version, possibly notify and write a new timestamp.
   */
  async notifyIfUpdateIsAvailable () {
    const log = this.log
    log.trace(`${this.constructor.name}.notifyIfUpdateIsAvailable()`)

    if (!this.latestVersionPromise) {
      log.trace(`${this.constructor.name}:` +
        ' latestVersionPromise not defined, update silently ignored.')
      // If the promise was not created, no action.
      return
    }

    let latestVersion
    try {
      // Actively await for the promise to complete.
      latestVersion = await this.latestVersionPromise

      log.trace(`${this.constructor.name}:` +
      ` latest ${this.packageName}${latestVersion}`)

      // Prevent further calls.
      this.latestVersionPromise = undefined

      if (semver.gt(latestVersion, this.packageVersion)) {
      // If versions differ, send the notification.
        this.sendNotification(latestVersion)
      }

      this.latestVersion = latestVersion

      if (this.private_.isRunningAsRoot) {
      // When running as root, skip writing the timestamp to avoid
      // later EACCES or EPERM. The effect is that the check will
      // be performed with each run.
      /* istanbul ignore next */
        if (os.platform() !== 'win32') {
          log.trace(`${this.constructor.name}:` +
          ` geteuid() ${process.geteuid()} != ${process.getuid()}`)
        } else {
        /* istanbul ignore next */
          log.trace(`${this.constructor.name}: running as administrator`)
        }
        return
      }
    } catch (err) {
      if (log.isDebug) {
        log.debug(err)
      } else {
        log.warning(err.message)
      }
      this.returnedError = err
    }

    try {
      await this.clearTimestamp()

      await this.createTimestamp()

      log.debug(`${this.constructor.name}.checkUpdate()` +
      ` timestamp ${this.timestampFileAbsolutePath} created`)
    } catch (err) /* istanbul ignore next */ {
      log.debug(err)
    }
  }

  /**
   * @summary Send the notification message.
   *
   * @param {String} latestVersion A string with the latest version,
   *  in semver format.
   * @return {undefined} Nothing.
   *
   * @description
   * The notification is sent by writing to the log.
   * This can be silenced by changing the log level.
   * To customise the behaviour, this function can be redefined
   * in a class derived from this one.
   *
   * @override
   */
  sendNotification (latestVersion) {
    const log = this.log
    log.trace(`${this.constructor.name}.sendNotification()`)

    const isGlobalStr = this.private_.isInstalledGlobally ? ' --global' : ''

    let msg = '\n'
    msg += `>>> New version ${this.packageVersion} -> `
    msg += `${latestVersion} available. <<<\n`
    msg += ">>> Run '"
    if (this.private_.isInstalledAsRoot) {
      msg += 'sudo '
    }
    msg += `npm install${isGlobalStr} ${this.packageName}' to update. <<<`

    log.info(msg)
  }

  /**
   * @summary Check if a subsequent test is too soon.
   *
   * @param {Number} ageMillis Number of milliseconds.
   * @returns {Boolean} True if there is no timestamp or if it is older
   *  than the given age.
   */
  async didIntervalExpire_ (ageMillis) {
    assert(ageMillis > 0)
    const log = this.log

    const stats = await this.readTimestamp()
    if (stats) {
      const millisDelta = Date.now() - stats.mtime
      if (millisDelta < ageMillis) {
        log.trace(`${this.constructor.name}: update timeout did not expire ` +
          `${Math.floor(millisDelta / 1000)} ` +
          `< ${Math.floor(ageMillis / 1000)}`)
        return false
      }
    } else {
      log.trace(`${this.constructor.name}:` +
        ' there is no previous update timestamp.')
    }

    return true
  }

  async readTimestamp () {
    const log = this.log

    try {
      const stats = await fsPromises.stat(this.timestampFileAbsolutePath)
      log.trace(`${this.constructor.name}.readTimestamp() ok`)
      return stats
    } catch (ex) {
      // File not found.
      // console.log(ex)
    }
    log.trace(`${this.constructor.name}.readTimestamp() failed`)
    return null
  }

  /**
   * @summary Remove the file used as timestamp
   *
   * @returns {undefined} Nothing.
   *
   * @description
   * The main reason this is a separate function is testability.
   */
  async clearTimestamp () {
    const log = this.log
    log.trace(`${this.constructor.name}.clearTimestamp()`)

    await del(this.timestampFileAbsolutePath, { force: true })
  }

  async createTimestamp () {
    const log = this.log
    log.trace(`${this.constructor.name}.createTimestamp()`)

    // Ensure the parent folder is present.
    await makeDir(path.dirname(this.timestampFileAbsolutePath))

    // Create an empty file; the content is ignored,
    // only the modified date is of interest.
    const fd = await fsPromises.open(this.timestampFileAbsolutePath, 'w')
    await fd.close()
  }
}

// ----------------------------------------------------------------------------
// Node.js specific export definitions.

// By default, `module.exports = {}`.
// The class is added as a property to this object.

module.exports.UpdateChecker = UpdateChecker

// In ES6, it would be:
// export class UpdateChecker { ... }
// ...
// import { UpdateChecker } from 'update-checker.js'

// ----------------------------------------------------------------------------
